lessons/Data-Science/CODAP/codap-composing/langs/en-us/index.adoc

```diff
@@ -0,0 +1,195 @@
+* Diagram their composition to make sense of the order of operations
+* Find bugs when Transformers are not composed in the correct order
+
+| Student-facing Lesson Goals
+|
+
+* Let's practice combining Transformers
+
+| Materials
+|[.materials-links]
+
+@material-links
+
+| Preparation
+|
+* This lesson requires each pair of students to create a cut-out set of "Transformer Cards". Have scissors available.
+|===
+
+== Creating Transformers (Review) @duration{25 minutes}
+
+=== Overview
+
+Students practice creating, naming, and saving all of the Trasnsformer types that they have discovered: Filter, Build Attribute, and Transform Attribute.
+
+=== Launch
+
+One of the perks of using a tool like CODAP is that it invites experimentation, prompting us to play with a variety of inputs until we have designed exactly the Transformer we need.
+
+Sometimes, however, relying on experimentation can limit our ability to _plan_ our data transformations. Familiarizing ourselves with exactly which inputs each Transformer requires can help us to efficiently transform data, without relying on guessing and checking every time.
+
+[.lesson-instruction]
+--
+- Each Transformer we've worked with consumes a different amount of arguments, represented in the table below.
+
+[cols= "10,10,10",options="header"]
+|===
+| Filter
+| Build Attribute
+| Transform Attribute
+
+| 1 Transformer Type
+
+2 Dataset
+
+3 Expression
+
+| 1 Transformer Type
+
+2 Dataset
+
+3 Name of New Attribute
+
+4 Expression
+
+| 1 Transformer Type
+
+2 Dataset
+
+3 Attribute to Transform
+
+4 New Name for Transformed Attribute
+
+5 Expression
+
+|===
+
+- How many arguments does Filter require? How about Build Attribute? Transform Attribute?
+** _Filter requires 2 arguments. Build Attribute requires 3 arguments. Transform Attribute requires 4 arguments._
+- Why does each Transformer require a different number of arguments?
+** _All Transformeres require a Transformer type and a dataset. Beyond that, Filter only needs a Boolean to know which rows to include and exclude from the new dataset. Build and Transform both require a new attribute name. Transform also needs to know which attribute is being transformed._
+--
+
+=== Investigate
+
+In the subsequent activities, "Transformer Cards" are a tool for students to think about what specific arguments each Transformer requires. These cards will later support students in composing Transformers.
+
+Let's practice creating Transformer Cards to represent various Transformers.
+
+[.lesson-instruction]
+- Turn to @printable-exercise{create-transformer-cards.adoc}.
+- Respond to the prompts, drawing Circles of Evaluation that represent the transformation needed to create the desired table.
+- In the last column, give each Transformer a useful and descriptive name.
+- Now, complete the handout @handout{create-transformer-cards-2.adoc}. On this page, you will need to determine how many arguments each Transformer requires before creating the appropriate card.
```

Comment:
What is the learning outcome here? That kids memorize how many arguments each transformer takes? I might need to be convinced that's a worthy goal. It feels like it's not contributing to the main objective of the lesson, and is instead taking up time and brainpower that should be used elsewhere.

Reply:
I'm cool with that. I was trying to avoid having students rely on three separate templates, but I had doubts myself here. I have no reservations putting the prompts in!